### PR TITLE
Add support for portenta-c33

### PR DESCRIPTION
**Pull request by MatteoPologruto**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,18 +6,23 @@ require (
 	github.com/arduino/arduino-cli v0.0.0-20230724101917-62bb4c47cb2b
 	github.com/arduino/fwuploader-plugin-helper v1.0.0
 	github.com/arduino/go-paths-helper v1.9.1
+	github.com/arduino/uno-r4-wifi-fwuploader-plugin v0.0.0-20230719120105-af756953af1e
```
Comment (umbynos) [inline go.mod]: why are you importing another plugin?

File: sketches/compile.sh
Comment (umbynos) [inline sketches/compile.sh:1]: create a task for that

File: main.go [hunk 2]
```diff
@@ -141,7 +148,7 @@ func (d *portentaC33Plugin) UploadCertificate(portAddress, fqbn string, certific
 	if err != nil {
 		return err
 	}
-	if err := connection.SetReadTimeout(5 * time.Second); err != nil {
+	if err := connection.SetReadTimeout(100 * time.Second); err != nil {
```
Comment (umbynos) [inline main.go]: Isn't 100 seconds a lot?

**Review by umbynos — DISMISSED** [review]

File: README.md [hunk 3]
```diff
@@ -1,3 +1,58 @@
...
+### Requirements
+
+1. [Go](https://go.dev/) version 1.20 or later
+1. [Task](https://taskfile.dev/) to help you run the most common tasks from the command line
+1. The [portenta c33](https://docs.arduino.cc/hardware/portenta-c33) board to test the core parts.
+
+## Development
+
+When running only the plugin without the fwuploader, the required tools are downloaded by the fwuploader. If you run only the plugin, you must provide them by hand.
+Therefore be sure to place the `esptool` and `dfu-util` binaries in the correct folders like the following:
```
Comment (cmaglie) [inline README.md]:
```suggestion
When running the plugin inside the fwuploader, the required tools are downloaded by the fwuploader. If you run only the plugin, you must provide them by hand.
Therefore be sure to place the `esptool` and `dfu-util` binaries in the correct folders like the following:
```

File: main.go [hunk 4]
```diff
@@ -1,4 +1,278 @@
...
+		return nil, err
+	}
+	defer port.Close()
+
+	var version string
+	scanner := bufio.NewScanner(port)
+	for scanner.Scan() {
+		version = scanner.Text()
+		break
+	}
```
Comment (cmaglie) [inline main.go]: Uhm... this loop may block forever if the sketch crashes, also the serial port does not behave nicely with `bufio.Scanner`.

I'd change this with just:
```go
// wait 1 second to allow the sketch to fill the buffer with the version string
time.Sleep(time.Second)

buff := make([]byte, 30)
port.SetReadTimeout(time.Second)
n, err := port.Read(buff)
if err != nil || n==0 { .... }
version = strings.TrimSpace(string(buff[:n]))
```

File: main.go [hunk 5]
```diff
@@ -1,4 +1,278 @@
...
+
+	return semver.ParseRelaxed(version), nil
+}
+
+func (d *portentaC33Plugin) uploadSketch(portAddress *string, feedback *helper.PluginFeedback, sketch *paths.Path) error {
```
Comment (cmaglie) [inline main.go]: I'd return the new port instead of changing the `portAddress` in place.

File: ymodem/ymodem.go [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//go:build !darwin
+
+package ymodem
```
Comment (cmaglie) [inline ymodem/ymodem.go]: Did you implement `ymodem` here?

Why the darwin implementation is different?

**Review by cmaglie — CHANGES_REQUESTED** [review]

File: ymodem/ymodem.go (hunk 6 above)
Comment (alessio-perugini) [reply ymodem/ymodem.go]: @facchinm

File: go.mod (hunk 1 above)
Comment (alessio-perugini) [reply go.mod]: We're using the `serial` pkg. Instead of copy-pasting the folder

File: ymodem/ymodem.go (hunk 6 above)
Comment (alessio-perugini) [reply ymodem/ymodem.go]: We're using our fork at the moment. https://github.com/arduino/go-xmodem

**Review by cmaglie — APPROVED** [review]

**Pull request merged**